### PR TITLE
Refactor FormulaCop as an abstract class

### PR DESCRIPTION
**Pull request by dduugg**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
Resolves the TODO that I introduced in https://github.com/Homebrew/brew/pull/14649/files#diff-67dd5bf1d7b1a755e268d61fa130a041ce6b42cf4c6587dc6ef2bb7bdd3b486fR51 by refactoring `FormulaCop` to be a mixin. (This was intended to be an abstract base class, but due to how rubocop uses reflection on subclasses of `Rubocop::Cop::Cop`, it was considered a cop itself.)

@Bo98 mentioned looking into `Force`, but this seems to be a lightly [deprecated](https://github.com/rubocop/rubocop/discussions/9255) (or at least "never fledged" out) concept.

There is some tangential cleanup that I'll clarify in comments, but lmk if you'd prefer a more focused PR instead.

### CODE REVIEW

File: Library/Homebrew/rubocops/all.rb [hunk 1]
```diff
@@ -13,29 +13,30 @@
 end
 
 require_relative "io_read"
+require_relative "move_to_extend_os"
 require_relative "shell_commands"
-require_relative "platform"
 
-require_relative "formula_desc"
+# formula audit cops
+require_relative "bottle"
```
Comment (dduugg) [inline Library/Homebrew/rubocops/all.rb:20]: I sorted the requires in each group, and added a comment clarifying with this group represents

File: Library/Homebrew/rubocops/desc.rb [hunk 2]
```diff
@@ -1,15 +1,16 @@
...
+require "rubocops/extend/formula_cop"
 require "rubocops/shared/desc_helper"
 
 module RuboCop
   module Cop
     module FormulaAudit
       # This cop audits `desc` in formulae.
       # See the {DescHelper} module for details of the checks.
-      class Desc < FormulaCop
+      class Desc < Base
```
Comment (dduugg) [inline Library/Homebrew/rubocops/desc.rb:12]: I renamed this file to match the class name

File: Library/Homebrew/rubocops/extend/formula_cop.rb [hunk 3]
```diff
@@ -5,10 +5,11 @@
 
 module RuboCop
   module Cop
-    # Superclass for all formula cops.
+    # Mixin for all formula cops.
     #
     # @api private
-    class FormulaCop < Base
+    module FormulaCop
```
Comment (dduugg) [inline Library/Homebrew/rubocops/extend/formula_cop.rb:11]: I renamed this file to match the class name

File: Library/Homebrew/rubocops/extend/formula_cop.rb [hunk 4]
```diff
@@ -5,10 +5,11 @@
...
+      extend RuboCop::AST::NodePattern::Macros
```
Comment (dduugg) [inline Library/Homebrew/rubocops/extend/formula_cop.rb:12]: This is necessary to support `def_node_search` below

File: Library/Homebrew/rubocops/extend/formula_cop.rb [hunk 5]
```diff
@@ -67,11 +68,7 @@ def depends_on?(dependency_name, *types)
       # Returns true if given dependency name and dependency type exist in given dependency method call node.
       # TODO: Add case where key of hash is an array
       def depends_on_name_type?(node, name = nil, type = :required)
-        name_match = if name
-          false
-        else
-          true # Match only by type when name is nil
-        end
+        name_match = !name # Match only by type when name is nil
```
Comment (dduugg) [inline Library/Homebrew/rubocops/extend/formula_cop.rb:71]: Couldn't help simplifying this expression while i was here 🤷

File: Library/Homebrew/rubocops/all.rb [hunk 6]
```diff
@@ -13,29 +13,30 @@
 end
 
 require_relative "io_read"
+require_relative "move_to_extend_os"
```
Comment (dduugg) [inline Library/Homebrew/rubocops/all.rb:16]: This is renamed from `platform` to match the class/cop name

**Comment by Bo98** [comment]
> @Bo98 mentioned looking into `Force`, but this seems to be a lightly [deprecated](https://github.com/rubocop/rubocop/discussions/9255) (or at least "never fledged" out) concept.

More the latter as RuboCop itself still uses them for performace reasons. I equally saw performance improvements of about 25% but that perhaps wasn't significant enough.

**Review by MikeMcQuaid — APPROVED** [review]
Looks good to me, nice work!

**Comment by dduugg** [comment]
It turns out that Rubocop has a DSL to label abstract cop classes: https://github.com/rubocop/rubocop/blob/2f08053/lib/rubocop/cop/base.rb#L76-L79

I'm inclined to redo this and go that route instead, which do folks prefer?

**Comment by Bo98** [comment]
> It turns out that Rubocop has a DSL to label abstract cop classes: https://github.com/rubocop/rubocop/blob/2f08053/lib/rubocop/cop/base.rb#L76-L79
> 
> I'm inclined to redo this and go that route instead, which do folks prefer?

I think that route makes sense yeah.

**Comment by dduugg** [comment]
> > It turns out that Rubocop has a DSL to label abstract cop classes: https://github.com/rubocop/rubocop/blob/2f08053/lib/rubocop/cop/base.rb#L76-L79
> > I'm inclined to redo this and go that route instead, which do folks prefer?
> 
> I think that route makes sense yeah.

Done! The diff is much smaller now 😸

**Review by Bo98 — APPROVED** [review]
Yeah, looks better!

**Review by MikeMcQuaid — APPROVED** [review]
Thanks again @dduugg!

**Review by Rylan12 — APPROVED** [review]

**Pull request merged**